test/BasicVaultTest.t.sol

```diff
@@ -0,0 +1,164 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.13;
+
+import "forge-std/Test.sol";
+import "forge-std/console2.sol";
+
+import 'src/BasicVault.sol';
+import 'test/WMDTokenWithFailedTransfers.sol';
+
+abstract contract StateZero is Test {
+        
+    BasicVault public vault;
+    WMDTokenWithFailedTransfers public wmd;
+    address user;
+
+    event Deposited(address indexed from, uint amount);
+    event Withdrawal(address indexed from, uint amount);
+
+    function setUp() public virtual {
+        wmd = new WMDTokenWithFailedTransfers();
+        vault = new BasicVault(wmd);
+
+        user = address(1);
+        vm.label(user, "user");
+    }
+}
+
+contract StateZeroTest is StateZero {
+    
+    function testGetUserBalance() public {
+        console2.log("Check User has zero tokens in vault");
+        vm.prank(user);
+        uint balance = vault.balances(user);
+        assertTrue(balance == 0);
+    }
+
+    function testUserCannotWithdraw(uint amount) public {
+        console2.log("User cannot withdraw with no balance");
+        vm.assume(amount > 0);
```

Comment:
Alberto said he wanted to see this revert because of the underflow, not because of the erc20 balance.  so u can:

1. make sure the user has an balance of the token
2. change this to vm.assume`(amount > 0 && amount < wmd.balanceOf(user))`
3. instead of expectRevert -- use forge-std standard errors https://book.getfoundry.sh/reference/forge-std/std-errors.html

Reply:
The takeaway here is that: 
1) the user has some wmd tokens.
2) he will attempt to withdraw some tokens within the range (0, amount), from the vault
3) since he does not have any deposits to begin with, we run into an underflow error, resulting from:
       ` balances[msg.sender] -= amount;`
as this is going to work out to be subtracting a positive figure (amount) from 0 -> hence underflow.

Reply:
also, you could go w just one vm.prank instead of using startPrank stopPrank since you don't need the wmd.mint to be done by user and all the other calls are cheats.  so u could just do:
```
vm.prank(user);
vault.withdrwaw(...
```
 and be done with it

Reply:
Follow-up question, as this test function requires the user to have minted tokens, would it not be better placed in StateMintedTest, as opposed to StateZero (where it currently resides).

Reply:
one more thing - normally u would do the wmd.mint in the setup of this fn, but not terrible for a one-off.  

but it's kind of weird to set the minting amount to exactly the amount as the fuzz param.  i would prob just pick a fixed number like 1_000_000 * 1e18 (or whatever the decimals are)

Reply:
> Follow-up question, as this test function requires the user to have minted tokens, would it not be better placed in StateMintedTest, as opposed to StateZero (where it currently resides).

Yeah this is much preferable. I usually have a `uint256 public userInitialBalanceTokenX;`  in the ZeroState contract and then mint the tokens in the ZeroState setup()

Reply:
You're right, else, I get [FAIL. Reason: Too many global rejects], when minting exactly the amount as the fuzz param